### PR TITLE
Custom message parsing issue #73

### PR DESCRIPTION
**Pull request by yavosh**
allow custom parser as per issue #73

### CODE REVIEW

File: parser.go [hunk 1]
```diff
@@ -5,13 +5,18 @@ import (
 	"strconv"
 )
 
-// parser provides a simple way of accessing and parsing
+// Parser provides a simple way of accessing and parsing
 // sentence fields
 type parser struct {
```
Comment (adrianmo) [inline parser.go:10]: Since we are exposing the parser to external users we should now export it.

The linter is also complaining about it.
```
parser.go:16:32: exported func NewParser returns unexported type *nmea.parser, which can be annoying to use
```

File: sentence.go [hunk 2]
```diff
@@ -19,6 +19,56 @@ const (
...
+		TypeRTE: func(s BaseSentence) (Sentence, error) {
+			return newRTE(s)
+		},
+		TypeVHW: func(s BaseSentence) (Sentence, error) {
+			return newVHW(s)
+		},
+	}
+)
+
+type parserCallbackType func(BaseSentence) (Sentence, error)
```
Comment (adrianmo) [inline sentence.go:70]: I suggest renaming and exporting `parserCallbackType` to `ParserCallback` and then use it in the `MustRegisterParser` and `RegisterParser` functions.

File: sentence.go [hunk 3]
```diff
@@ -110,48 +160,45 @@ func Checksum(s string) string {
 	return fmt.Sprintf("%02X", checksum)
 }
 
+// MustRegisterParser register a custom parser or panic
+func MustRegisterParser(t string, parser func(BaseSentence) (Sentence, error)) {
```
Comment (adrianmo) [inline sentence.go:164]: Change to `func MustRegisterParser(t string, parser ParserCallback)`.

File: sentence.go [hunk 4]
```diff
@@ -110,48 +160,45 @@ func Checksum(s string) string {
...
+	if err := RegisterParser(t, parser); err != nil {
+		panic(err)
+	}
+}
+
+// RegisterParser register a custom parser
+func RegisterParser(t string, parser func(BaseSentence) (Sentence, error)) error {
```
Comment (adrianmo) [inline sentence.go:171]: Change to `func RegisterParser(t string, parser ParserCallback) error`

**Comment by adrianmo** [comment]
@icholy I would like you to review the changes when you have a chance. Thanks!

File: sentence_customparser_test.go [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+func TestWillReturnErrorOnDuplicateRegistration(t *testing.T) {
+	err := RegisterParser("XXX", func(s BaseSentence) (Sentence, error) {
+		return BaseSentence{}, nil
+	})
+	assert.NoError(t, err)
+
+	err = RegisterParser("XXX", func(s BaseSentence) (Sentence, error) {
+		return BaseSentence{}, nil
+	})
+	assert.Error(t, err)
```
Comment (adrianmo) [inline sentence_customparser_test.go:138]: We could also add a test for how MustRegisterParser panics if on duplicate registration.

e.g.:
```
	f := func() {
		MustRegisterParser("XXX", func(s BaseSentence) (Sentence, error) {
			return BaseSentence{}, nil
		})
	}
	assert.Panics(t, f)
```

Comment (yavosh) [reply sentence_customparser_test.go:138]: ok

File: parser.go (hunk 1 above)
Comment (yavosh) [reply parser.go]: wasnt sure if it was like this on purpose, done

**Comment by yavosh** [comment]
please take an other look.

File: sentence_customparser_test.go [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+}
+
+func TestWillPanicOnDuplicateMustRegister(t *testing.T) {
+	MustRegisterParser("AAA", func(s BaseSentence) (Sentence, error) {
+		return BaseSentence{}, nil
+	})
+
+	assert.PanicsWithError(t, "nmea: parser for prefix 'AAA' already exists", func() {
```
Comment (adrianmo) [inline sentence_customparser_test.go:146]: Can you update the testify dependency to use the latest v1.5.1 in the `go.mod` file?

The current version (v1.2.1) doesn't have the `PanicsWithError` function. The CI doesn't fail because we are doing `go get ./...` instead of `go mod download`. I'll fix that in a separate PR.

File: sentence.go [hunk 7]
```diff
@@ -19,6 +19,66 @@ const (
...
+			return newDBS(s)
+		},
+		TypeDPT: func(s BaseSentence) (Sentence, error) {
+			return newDPT(s)
+		},
+	}
+)
+
+// ParserCallback callback used to parse specific sentence variants
+type ParserCallback func(BaseSentence) (Sentence, error)
```
Comment (icholy) [inline sentence.go:80]: The convention in Go is to use `Func` instead of `Callback` in types for functions. I'd prefer `ParseFunc`.

File: sentence.go [hunk 8]
```diff
@@ -19,6 +19,66 @@ const (
 	ChecksumSep = "*"
 )
 
+var (
+	customParsers          = map[string]ParserCallback{}
+	defaultSentenceParsers = map[string]ParserCallback{
```
Comment (icholy) [inline sentence.go:24]: Call this `defaultParsers`.

File: sentence.go [hunk 9]
```diff
@@ -19,6 +19,66 @@ const (
 	ChecksumSep = "*"
 )
 
+var (
+	customParsers          = map[string]ParserCallback{}
```
Comment (icholy) [inline sentence.go:23]: Protect access to `customParsers` with a `sync.Mutex`. Call it `customParsersMu`.

File: sentence.go [hunk 10]
```diff
@@ -110,54 +170,45 @@ func Checksum(s string) string {
...
-		case TypeDPT:
-			return newDPT(s)
-		case TypeDBT:
-			return newDBT(s)
-		case TypeDBS:
-			return newDBS(s)
+
+		if parserCallback, ok := defaultSentenceParsers[s.Type]; ok {
+			return parserCallback(s)
 		}
```
Comment (icholy) [inline sentence.go:212]: Does this change need to happen?

File: sentence.go [hunk 11]
```diff
@@ -110,54 +170,45 @@ func Checksum(s string) string {
...
+func MustRegisterParser(t string, parser ParserCallback) {
+	if err := RegisterParser(t, parser); err != nil {
+		panic(err)
+	}
+}
+
+// RegisterParser register a custom parser
+func RegisterParser(t string, parser ParserCallback) error {
+	if _, ok := customParsers[t]; ok {
+		return fmt.Errorf("nmea: parser for prefix '%s' already exists", t)
```
Comment (icholy) [inline sentence.go:183]: Use `%q` instead of `'%s'`.

File: sentence.go [hunk 12]
```diff
@@ -110,54 +170,45 @@ func Checksum(s string) string {
...
 
+// MustRegisterParser register a custom parser or panic
+func MustRegisterParser(t string, parser ParserCallback) {
+	if err := RegisterParser(t, parser); err != nil {
+		panic(err)
+	}
+}
+
+// RegisterParser register a custom parser
+func RegisterParser(t string, parser ParserCallback) error {
```
Comment (icholy) [inline sentence.go:181]: `t` is too ambiguous. It could be referring to the type or talker.

File: sentence.go [hunk 13]
```diff
@@ -110,54 +170,45 @@ func Checksum(s string) string {
...
+
 // Parse parses the given string into the correct sentence type.
 func Parse(raw string) (Sentence, error) {
 	s, err := parseSentence(raw)
 	if err != nil {
 		return nil, err
 	}
+
+	// Custom parser allow overriding of existing parsers
+	if customParserCallback, ok := customParsers[s.Type]; ok {
```
Comment (icholy) [inline sentence.go:198]: Call this variable `parser`.

File: sentence.go [hunk 14]
```diff
@@ -110,54 +170,45 @@ func Checksum(s string) string {
...
-		case TypeVHW:
-			return newVHW(s)
-		case TypeDPT:
-			return newDPT(s)
-		case TypeDBT:
-			return newDBT(s)
-		case TypeDBS:
-			return newDBS(s)
+
+		if parserCallback, ok := defaultSentenceParsers[s.Type]; ok {
```
Comment (icholy) [inline sentence.go:210]: I think custom parsers should be forced to match the entire message prefix, talker + type.

Comment (icholy) [inline sentence.go:210]: I think custom parsers should be forced to match the entire message prefix, talker + type.

File: sentence.go (hunk 13 above)
Comment (icholy) [reply sentence.go:198]: I think custom parsers should be forced to match the entire message prefix, talker + type.

Comment (yavosh) [reply sentence.go:198]: I disagree, in my experience different talkers can provide the same sentence type. I do see a case where you might want to match on the whole prefix though, perhaps this can be a future task.

File: sentence.go (hunk 9 above)
Comment (yavosh) [reply sentence.go]: ok

File: sentence.go (hunk 10 above)
Comment (yavosh) [reply sentence.go]: Made sence to me at the time, I suppose not.

**Comment by yavosh** [comment]
Added example

File: README.md [hunk 15]
```diff
@@ -88,6 +88,79 @@ Date: 13/06/94
...
+
+Output:
+
+```
+$ go run main/main.go
+
+Raw sentence: $AAXYZ,220516,A,23,5133.82,W*42
+Time: 22:05:16.0000
+Label: A
+Vounter: 23
```
Comment (adrianmo) [inline README.md:160]: Typo -> `Counter`.
